lib/devise/models/confirmable.rb

```diff
@@ -349,6 +349,17 @@ def send_confirmation_instructions(attributes={})
         # Options must have the confirmation_token
         def confirm_by_token(confirmation_token)
           confirmable = find_first_by_auth_conditions(confirmation_token: confirmation_token)
+
+          # When the `confirmation_token` parameter is blank, if there are any users with a blank
+          # `confirmation_token` in the database, the first one would be confirmed here.
+          # The error is being manually added here to ensure no users are confirmed by mistake.
+          # This was done in the model for convenience, since validation errors are automatically
+          # displayed in the view.
+          if confirmable && confirmation_token.blank?
```

Comment:
As written, this code may leak information about the internal state of the database (that some records have blank tokens). Could we check for presence of confirmation_token before the call to `find_first_by_auth_conditions`?

Reply:
> this code may leak information about the internal state of the database (that some records have blank tokens)

I'm not sure about that. It's the same response as when there are no records with blank tokens. Do you think this is still leaking information?

Reply:
If that's the case then I guess it's probably ok for now, but if the `find_or_initialize_with_error_by` or `token_generator` methods change, it could unintentionally create an information leakage issue here. I think it'd be safer to just return early (without making any queries) if we detect that the token is blank.

```ruby
if confirmation_token.blank?
  confirmable = new
  confirmable.errors.add(:confirmation_token, :blank)
  return confirmable
end
```

Reply:
> but if the find_or_initialize_with_error_by or token_generator methods change, it could unintentionally create an information leakage issue here

Do you have examples of which kind of change could cause leaks?

Reply:
> > but if the find_or_initialize_with_error_by or token_generator methods change, it could unintentionally create an information leakage issue here
> 
> Do you have examples of which kind of changes could cause leaks?

Any change that would result in a different response being returned when the token is blank and there is no record in the database.

Examples:
- `TokenGenerator#digest` changes to return a digest even when `value` is blank. This would result in a record being initialized and returned with error set to `:invalid` instead of `:blank`.
- `find_or_initialize_with_error_by` changes to always use the provided `error` argument instead of special-casing blank errors.

Reply:
Ok, get it. Both of these cases would be caught by our integration tests, so I guess this implementation is fine.
But since we already know there's no need to perform the query I guess it's better to return early as you mentioned. It also shows better the intention behind the code, which is to validate whether the parameter is present or not.

Reply:
I've added the changes, thanks for the review!